File: src/engine/host.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"quackitect/engine/internal/quiet"
)

// WHERE THIS BOX IS, READ OFF ONE TABLE.
//
// The cage asks util/cage/hosts.json through util/cage/host.mjs, and this is
// the engine's reading of the same file. Two readers of one table can only
// disagree by reading it wrong, which is a smaller thing than two tables.
//
// A cloud box is one nobody sits beside. The tree was cloned cold, nothing in
// .bin travelled, and the only way a person learns what happened there is
// whatever the agent carries out. So the record says where it was written,
// and the diagnosis below is how the box explains itself.

// Host is where this box is, and which variable said so.
type Host struct {
	Cloud   bool   `json:"cloud"`
	Harness string `json:"harness"`
	Because string `json:"because"`
	Says    string `json:"says"`
}

type hostTable struct {
	Clouds []struct {
		Harness string `json:"harness"`
		Env     string `json:"env"`
		Is      string `json:"is"`
		Says    string `json:"says"`
	} `json:"clouds"`
}

// TheHost answers off the table under the method root. A table that cannot
// be read answers a desk, and says why, because the worse mistake is a box
// that believes it is in the cloud and starts fetching branches on a desk.
func TheHost(method string) Host {
	b, err := os.ReadFile(filepath.Join(method, "util", "cage", "hosts.json"))
	if err != nil {
		return Host{Because: "util/cage/hosts.json could not be read: " + err.Error(),
			Says: "a box with a person beside it"}
	}
	var t hostTable
	if err := json.Unmarshal(b, &t); err != nil {
		return Host{Because: "util/cage/hosts.json does not read: " + err.Error(),
			Says: "a box with a person beside it"}
	}
	named := ""
	for i, c := range t.Clouds {
		if os.Getenv(c.Env) == c.Is {
			return Host{Cloud: true, Harness: c.Harness, Because: c.Env + "=" + c.Is, Says: c.Says}
		}
		if i > 0 {
			named += ", "
		}
		named += c.Env
	}
	return Host{Because: "none of " + named + " is set", Says: "a box with a person beside it"}
}

// Diagnose writes a diagnosis of this box and prints it, and answers the exit
// code of the program that did.
//
// THE DIAGNOSIS IS MEASURED AND NOT REASONED. An agent on a lane-less cloud box
// read the lane's source and explained a failure the source no longer had, and
// a fix was built on that reading. So the script asks the box what is there:
// the commit against origin, the built programs against their source, the
// engine's own answer, the lane's log under .se/lane.out, and the network. It
// lives in util/cage/diagnose.mjs so a tree with nothing built can run it
// with node alone, and this flag is the same call through the engine, which is
// the one program the write gate lets a shell run.
func Diagnose(r Roots) int {
	script := filepath.Join(r.Method, "util", "cage", "diagnose.mjs")
	cmd := quiet.Quietly(exec.Command("node", script, "--method", r.Method, "--work", r.Work))
	cmd.Dir = r.Work
	cmd.Stdout, cmd.Stderr, cmd.Stdin = os.Stdout, os.Stderr, nil
	if err := cmd.Run(); err != nil {
		if exit, ok := err.(*exec.ExitError); ok {
			return exit.ExitCode()
		}
		fmt.Fprintln(os.Stderr, "quackitect: the diagnosis could not run:", err)
		return 1
	}
	return 0
}

File: util/cage/host.mjs
// WHERE THIS BOX IS, ANSWERED OFF ONE TABLE.
//
// A cloud box clones the tree cold, has no person beside it, and runs a
// harness that names its cloud with a variable of its own. The cage has to
// behave differently there, and it asked one harness's variable in one place
// and nothing in the others. So the table is util/cage/hosts.json, and this is
// the door the cage reads it through. The engine reads the same file in
// src/engine/host.go, so the two never disagree about where they are.
//
// WHY node AND NOT sh: see util/cage/mcp-lane.mjs.
//
//   node util/cage/host.mjs            prints the answer as JSON
//   node util/cage/host.mjs --cloud    exits 0 on a cloud box, 1 anywhere else
//   node util/cage/host.mjs --say      prints one line a person can read
import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const table = JSON.parse(readFileSync(join(dirname(fileURLToPath(import.meta.url)), "hosts.json"), "utf8"));

// theHost answers where this box is: cloud or not, which harness said so, and
// the variable it said it with.
export function theHost(env = process.env) {
  for (const c of table.clouds) {
    if ((env[c.env] ?? "") === c.is) {
      return { cloud: true, harness: c.harness, because: c.env + "=" + c.is, says: c.says };
    }
  }
  return {
    cloud: false, harness: "",
    because: "none of " + table.clouds.map((c) => c.env).join(", ") + " is set",
    says: "a box with a person beside it",
  };
}

// Run as a program, this answers on the command line. Imported, it only exports.
if (process.argv[1] && pathToFileURL(resolve(process.argv[1])).href === import.meta.url) {
  const host = theHost();
  const flag = process.argv[2] ?? "";
  if (flag === "--cloud") {
    process.exit(host.cloud ? 0 : 1);
  }
  if (flag === "--say") {
    console.log(host.says + " (" + host.because + ")");
  } else {
    console.log(JSON.stringify(host, null, 2));
  }
}
